File: src/main/java/no/nav/veilarbaktivitet/aktivitetskort/AktivitetskortCompareUtil.java
package no.nav.veilarbaktivitet.aktivitetskort;

import com.fasterxml.jackson.core.JsonProcessingException;
import lombok.extern.log4j.Log4j;
import lombok.extern.slf4j.Slf4j;
import no.nav.common.json.JsonUtils;
import no.nav.veilarbaktivitet.aktivitet.domain.AktivitetData;

@Slf4j
public class AktivitetskortCompareUtil {
    private AktivitetskortCompareUtil() {}

    public static boolean erFaktiskOppdatert(AktivitetData innkommende, AktivitetData eksisterende) {
        var mapper = JsonUtils.getMapper();
        var eksisterendeMedRelevanteFelter = eksisterende
                .withId(null)
                .withForhaandsorientering(null)
                .withOppfolgingsperiodeId(null)
                .withVersjon(null)
                .withTransaksjonsType(null)
                .withStatus(null);
        // Innkommende vil aldri ha interne tekniske ider, transaksjonstype eller oppfølgingsperiode
        var innkommendeMedRelevanteFelter = innkommende
                .withId(null)
                .withForhaandsorientering(null)
                .withOppfolgingsperiodeId(null)
                .withVersjon(null)
                .withTransaksjonsType(null)
                .withStatus(null);
        try {
            return !mapper.writeValueAsString(eksisterendeMedRelevanteFelter)
                .equals(mapper.writeValueAsString(innkommendeMedRelevanteFelter));
            } catch (JsonProcessingException e) {
            throw new IllegalStateException("Kunne ikke parse aktiviteter for sammenligning", e);
        }

    }

}

File: src/main/kotlin/no/nav/veilarbaktivitet/aktivitetskort/AktivitetskortCompareUtil.kt
package no.nav.veilarbaktivitet.aktivitetskort

import lombok.extern.slf4j.Slf4j
import no.nav.veilarbaktivitet.aktivitet.domain.AktivitetData
import no.nav.veilarbaktivitet.aktivitet.domain.aktiviteter.AktivitetMuterbareFelter
import no.nav.veilarbaktivitet.aktivitet.domain.aktiviteter.Eksternaktivitet

@Slf4j
object AktivitetskortCompareUtil {
    fun erFaktiskOppdatert(innkommende: Eksternaktivitet.Endre, eksisterende: AktivitetData): Boolean {
        val eksisterendeSomEndre = Eksternaktivitet.Endre(
            id = eksisterende.id,
            versjon = eksisterende.versjon,
            muterbareFelter = AktivitetMuterbareFelter(
                tittel = eksisterende.tittel,
                beskrivelse = eksisterende.beskrivelse,
                fraDato = eksisterende.fraDato,
                tilDato = eksisterende.tilDato,
                lenke = eksisterende.lenke,
            ),
            sporing = innkommende.sporing, // Ikke sammenlignet
            eksternAktivitetData = eksisterende.eksternAktivitetData,
            erAvtalt = eksisterende.isAvtalt, // Ikke sammenlignet - håndteres i eget kall
            status = eksisterende.status // Ikke sammenlignet
        )

        // Sammenlign bare muterbare felter og ekstern aktivitet data
        val muterbareFelterEndret =
            innkommende.muterbareFelter.tittel != eksisterendeSomEndre.muterbareFelter.tittel
                || innkommende.muterbareFelter.beskrivelse != eksisterendeSomEndre.muterbareFelter.beskrivelse
                || innkommende.muterbareFelter.fraDato != eksisterendeSomEndre.muterbareFelter.fraDato
                || innkommende.muterbareFelter.tilDato != eksisterendeSomEndre.muterbareFelter.tilDato
                || innkommende.muterbareFelter.lenke != eksisterendeSomEndre.muterbareFelter.lenke

        val eksternDataEndret = innkommende.eksternAktivitetData != eksisterendeSomEndre.eksternAktivitetData

        return muterbareFelterEndret || eksternDataEndret
    }
}
